Modules/gcmodule.c

```diff
@@ -462,6 +462,43 @@ visit_decref(PyObject *op, void *parent)
     return 0;
 }
 
+int tupletraverse(PyTupleObject *o, visitproc visit, void *arg);
+int func_traverse(PyFunctionObject *f, visitproc visit, void *arg);
+int cell_traverse(PyCellObject *op, visitproc visit, void *arg);
+int dict_traverse(PyObject *op, visitproc visit, void *arg);
+int list_traverse(PyListObject *o, visitproc visit, void *arg);
+
+/* Optimization:
+ * Calling inlined_tp_traverse(op, proc, c) is equivalent to calling
+ * Py_TYPE(op)->tp_traverse(op, proc, c), but is intended to be faster.  The
+ * approach is to check for common Python types and statically call their
+ * traverse functions (changing an indirect call to a static call), and by
+ * using a combination of the always_inline and flatten attributes to get the
+ * compiler to generate specializations of the traversal methods, getting rid
+ * of the indirect calls via Py_VISIT.
+ */
+__attribute__((always_inline)) __attribute__((flatten))
+static void
+inlined_tp_traverse(PyObject* op, visitproc proc, void* c)
+{
+    // This set of types and the order that they are checked was driven by data
+    // from a webserving macrobenchmark, but a more robust analysis and
+    // optimization should still be possible.
+    if (Py_TYPE(op) == &PyTuple_Type) {
```

Comment:
This is a bit of a gamble here because for types that are not these this is going to introduce N branch mispredictions for instance. Indeed, I did a quick benchmarks creating a lot of instances from extension classes and running a GC collection and the collection was 12% slower.

I agree that in most situations these types are going to be common but this is not a clear win in all cases and given that in macro benchmarks we are getting ~1% I am a bit hesitant

Reply:
We can easily measure the mispredictions using 'perf'